File: oscar-cp/src/test/java/oscar/cp/test/TestDeviation.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.test;


import junit.framework.TestCase;
import oscar.algo.Inconsistency;
import oscar.cp.constraints.Deviation;
import oscar.cp.core.variables.CPIntVar;
import oscar.cp.core.CPStore;



/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class TestDeviation extends TestCase {



    public TestDeviation(String name) {
        super(name);
    }
    	
	 /**
     * setUp() method that initializes common objects
     */
    protected void setUp() throws Exception {
        super.setUp();
    }

    /**
     * tearDown() method that cleanup the common objects
     */
    protected void tearDown() throws Exception {
        super.tearDown();
    }



    public void testDeviation1() throws Inconsistency {
        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	for (int i = 0; i < x.length; i++) {
			x[i] = CPIntVar.apply(cp,-2,2);
		}
    	CPIntVar nd = CPIntVar.apply(cp,0,0);
    	cp.post(new Deviation(x,0,nd));
        assertTrue(!cp.isFailed());
        for (int i = 0; i < x.length; i++) {
            assertTrue(x[i].isBound() && x[i].min() == 0);
        }
    }

    public void testDeviation2() throws Inconsistency{
        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	for (int i = 0; i < x.length; i++) {
			x[i] = CPIntVar.apply(cp,-2,2);
		}
    	CPIntVar nd = CPIntVar.apply(cp,0,6);
    	cp.post(new Deviation(x,1,nd));
        assertTrue(!cp.isFailed());

    }



    public void testDeviation3() throws Inconsistency{

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	x[0] = CPIntVar.apply(cp,3,7);
        x[1] = CPIntVar.apply(cp,0,5);
        x[2] = CPIntVar.apply(cp,5,6);
        x[3] = CPIntVar.apply(cp,5,7);

    	CPIntVar nd = CPIntVar.apply(cp,0,18);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[0].getMax(),5);

    }

    public void testDeviation4() throws Inconsistency{

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	x[0] = CPIntVar.apply(cp,3,7);
        x[1] = CPIntVar.apply(cp,0,5);
        x[2] = CPIntVar.apply(cp,5,6);
        x[3] = CPIntVar.apply(cp,5,7);

    	CPIntVar nd = CPIntVar.apply(cp,0,12);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[0].getMax(),4);

    }

    public void testDeviation5()  throws Inconsistency{

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	x[0] = CPIntVar.apply(cp,3,10);
        x[1] = CPIntVar.apply(cp,4,5);
        x[2] = CPIntVar.apply(cp,3,6);
        x[3] = CPIntVar.apply(cp,0,2);

    	CPIntVar nd = CPIntVar.apply(cp,0,45);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[0].getMin(),4);
        assertEquals(x[0].getMax(),9);

    }

    public void testDeviation6() throws Inconsistency {

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	x[0] = CPIntVar.apply(cp,3,10);
        x[1] = CPIntVar.apply(cp,4,5);
        x[2] = CPIntVar.apply(cp,3,6);
        x[3] = CPIntVar.apply(cp,0,2);

    	CPIntVar nd = CPIntVar.apply(cp,0,22);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[0].getMin(),4);
        assertEquals(x[0].getMax(),7);

    }

    public void testDeviation7() throws Inconsistency {

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];
    	x[0] = CPIntVar.apply(cp,3,10);
        x[1] = CPIntVar.apply(cp,4,5);
        x[2] = CPIntVar.apply(cp,3,6);
        x[3] = CPIntVar.apply(cp,0,2);

    	CPIntVar nd = CPIntVar.apply(cp,0,30);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[0].getMin(),4);
        assertEquals(x[0].getMax(),8);

    }

    public void testDeviation8() throws Inconsistency {

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[4];

        x[0] = CPIntVar.apply(cp,4,5);
        x[1] = CPIntVar.apply(cp,3,6);
        x[2] = CPIntVar.apply(cp,0,2);
        x[3] = CPIntVar.apply(cp,3,10);

    	CPIntVar nd = CPIntVar.apply(cp,0,30);
    	cp.post(new Deviation(x,17,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[3].getMin(),4);
        assertEquals(x[3].getMax(),8);

    }

    public void testDeviation9() throws Inconsistency {

        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[6];

        x[0] = CPIntVar.apply(cp,11,16);
        x[1] = CPIntVar.apply(cp,9,11);
        x[2] = CPIntVar.apply(cp,12,14);
        x[3] = CPIntVar.apply(cp,13,14);
        x[4] = CPIntVar.apply(cp,10,12);
        x[5] = CPIntVar.apply(cp,12,15);


    	CPIntVar nd = CPIntVar.apply(cp,0,1000);
    	cp.post(new Deviation(x,74,nd));
        assertTrue(!cp.isFailed());
        assertEquals(nd.getMin(),24);
    }


  
    
    public void testDeviation11() throws Inconsistency{
        CPStore cp = new CPStore();
    	CPIntVar [] x = new CPIntVar[8];

        x[0] = CPIntVar.apply(cp,-27,-25);
        x[1] = CPIntVar.apply(cp,-27,-27);
		x[2] = CPIntVar.apply(cp,-27,-25);
        x[3] = CPIntVar.apply(cp,-27,-25);
		x[4] = CPIntVar.apply(cp,-30,-30);
        x[5] = CPIntVar.apply(cp,-27,-25);
		x[6] = CPIntVar.apply(cp,-27,-25);
        x[7] = CPIntVar.apply(cp,-27,-23);


    	CPIntVar nd = CPIntVar.apply(cp,0,75);
    	cp.post(new Deviation(x,-213,nd));
        assertTrue(!cp.isFailed());
        assertEquals(x[7].getMax(),-24);

    }
    
  public void testDeviation12() throws Inconsistency{
	CPStore cp = new CPStore();
	CPIntVar [] x = new CPIntVar[6];

	x[0] = CPIntVar.apply(cp,11,16);
	x[1] = CPIntVar.apply(cp,9,11);
	x[2] = CPIntVar.apply(cp,12,14);
	x[3] = CPIntVar.apply(cp,13,14);
	x[4] = CPIntVar.apply(cp,10,12);
	x[5] = CPIntVar.apply(cp,12,15);



	CPIntVar nd = CPIntVar.apply(cp,0,34);
	cp.post(new Deviation(x,74,nd));
	assertTrue(!cp.isFailed());
    assertEquals(x[0].getMax(),14);

 }
    
  public void testDeviation13() throws Inconsistency{
	CPStore cp = new CPStore();
	CPIntVar [] x = new CPIntVar[6];
	
	x[0] = CPIntVar.apply(cp,-14,-12);
	x[1] = CPIntVar.apply(cp,-11,-9);
	x[2] = CPIntVar.apply(cp,-14,-12);
	x[3] = CPIntVar.apply(cp,-14,-13);
	x[4] = CPIntVar.apply(cp,-12,-10);
	x[5] = CPIntVar.apply(cp,-14,-12);



	CPIntVar nd = CPIntVar.apply(cp,0,34);
	cp.post(new Deviation(x,-74,nd));
	assertTrue(!cp.isFailed());
    assertEquals(x[1].getMax(),-10);

 }
    

}

File: src/test/scala/oscar/cp/constraints/TestDeviation.scala
package oscar.cp.constraints

import oscar.cp.testUtils._

import oscar.cp._
import oscar.cp.constraints.Deviation
import oscar.cp.testUtils.TestSuite

class TestDeviation extends TestSuite {

  test("testDeviation1") {
    val cp = CPSolver()
    val x = Array.tabulate(4)(_ => CPIntVar(-2 to 2)(cp))
    val nd = CPIntVar(0 to 0)(cp)
    cp.add(new Deviation(x, 0, nd))
    cp.isFailed should be(false)
    for (i <- x.indices) {
      x(i).isBound should be(true)
      x(i).getMin should be(0)
    }
  }

  test("testDeviation2") {
    val cp = CPSolver()
    val x = Array.tabulate(4)(_ => CPIntVar(-2 to 2)(cp))
    val nd = CPIntVar(0 to 6)(cp)
    cp.add(new Deviation(x, 1, nd))
    cp.isFailed should be(false)
  }

  test("testDeviation3") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(3 to 7)(cp)
    x(1) = CPIntVar(0 to 5)(cp)
    x(2) = CPIntVar(5 to 6)(cp)
    x(3) = CPIntVar(5 to 7)(cp)

    val nd = CPIntVar(0 to 18)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(0).getMax should be(5)
  }

  test("testDeviation4") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(3 to 7)(cp)
    x(1) = CPIntVar(0 to 5)(cp)
    x(2) = CPIntVar(5 to 6)(cp)
    x(3) = CPIntVar(5 to 7)(cp)

    val nd = CPIntVar(0 to 12)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(0).getMax should be(4)
  }

  test("testDeviation5") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(3 to 10)(cp)
    x(1) = CPIntVar(4 to 5)(cp)
    x(2) = CPIntVar(3 to 6)(cp)
    x(3) = CPIntVar(0 to 2)(cp)

    val nd = CPIntVar(0 to 45)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(0).getMin should be(4)
    x(0).getMax should be(9)
  }

  test("testDeviation6") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(3 to 10)(cp)
    x(1) = CPIntVar(4 to 5)(cp)
    x(2) = CPIntVar(3 to 6)(cp)
    x(3) = CPIntVar(0 to 2)(cp)

    val nd = CPIntVar(0 to 22)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(0).getMin should be(4)
    x(0).getMax should be(7)
  }

  test("testDeviation7") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(3 to 10)(cp)
    x(1) = CPIntVar(4 to 5)(cp)
    x(2) = CPIntVar(3 to 6)(cp)
    x(3) = CPIntVar(0 to 2)(cp)

    val nd = CPIntVar(0 to 30)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(0).getMin should be(4)
    x(0).getMax should be(8)
  }

  test("testDeviation8") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](4)
    x(0) = CPIntVar(4 to 5)(cp)
    x(1) = CPIntVar(3 to 6)(cp)
    x(2) = CPIntVar(0 to 2)(cp)
    x(3) = CPIntVar(3 to 10)(cp)

    val nd = CPIntVar(0 to 30)(cp)
    cp.add(new Deviation(x, 17, nd))
    cp.isFailed should be(false)
    x(3).getMin should be(4)
    x(3).getMax should be(8)
  }

  test("testDeviation9") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](6)
    x(0) = CPIntVar(11 to 16)(cp)
    x(1) = CPIntVar(9 to 11)(cp)
    x(2) = CPIntVar(12 to 14)(cp)
    x(3) = CPIntVar(13 to 14)(cp)
    x(4) = CPIntVar(10 to 12)(cp)
    x(5) = CPIntVar(12 to 15)(cp)

    val nd = CPIntVar(0 to 1000)(cp)
    cp.add(new Deviation(x, 74, nd))
    cp.isFailed should be(false)
    nd.getMin should be(24)
  }

  test("testDeviation11") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](8)
    x(0) = CPIntVar(-27 to -25)(cp)
    x(1) = CPIntVar(-27 to -27)(cp)
    x(2) = CPIntVar(-27 to -25)(cp)
    x(3) = CPIntVar(-27 to -25)(cp)
    x(4) = CPIntVar(-30 to -30)(cp)
    x(5) = CPIntVar(-27 to -25)(cp)
    x(6) = CPIntVar(-27 to -25)(cp)
    x(7) = CPIntVar(-27 to -23)(cp)

    val nd = CPIntVar(0 to 75)(cp)
    cp.add(new Deviation(x, -213, nd))
    cp.isFailed should be(false)
    x(7).getMax should be(-24)
  }

  test("testDeviation12") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](6)
    x(0) = CPIntVar(11 to 16)(cp)
    x(1) = CPIntVar(9 to 11)(cp)
    x(2) = CPIntVar(12 to 14)(cp)
    x(3) = CPIntVar(13 to 14)(cp)
    x(4) = CPIntVar(10 to 12)(cp)
    x(5) = CPIntVar(12 to 15)(cp)

    val nd = CPIntVar(0 to 34)(cp)
    cp.add(new Deviation(x, 74, nd))
    cp.isFailed should be(false)
    x(0).getMax should be(14)
  }

  test("testDeviation13") {
    val cp = CPSolver()
    val x = new Array[CPIntVar](6)
    x(0) = CPIntVar(-14 to -12)(cp)
    x(1) = CPIntVar(-11 to -9)(cp)
    x(2) = CPIntVar(-14 to -12)(cp)
    x(3) = CPIntVar(-14 to -13)(cp)
    x(4) = CPIntVar(-12 to -10)(cp)
    x(5) = CPIntVar(-14 to -12)(cp)

    val nd = CPIntVar(0 to 34)(cp)
    cp.add(new Deviation(x, -74, nd))
    cp.isFailed should be(false)
    x(1).getMax should be(-10)
  }
}
